test/teleportation.slq

```diff
@@ -0,0 +1,35 @@
+// Function for alice to  operate on 'msg'
+def alice[n:!ℕ](msg:uint[n], qr:uint[n]):!uint[n]^2 {
+	// Perform a Bell measurement
+	for i in [0..n) {
+		if msg[i] { qr[i] := X(qr[i]); }
+		msg[i] := H(msg[i]);
+	}
+
+	return measure(msg, qr);
+}
+
+// Function for bob to recreate 'msg'
+def bob[n:!ℕ](qr:uint[n], aliceBits:!uint[n]^2):uint[n] {
+	// Conditionally perform bit-flip and/or phase-flip on bob's qureg
+	for i in [0..n) {
+		if aliceBits[0][i] { qr[i] := Z(qr[i]); }
+		if aliceBits[1][i] { qr[i] := X(qr[i]); }
+	}
+
+	return qr;
+}
+
+def createBellPair[n:!ℕ]():uint[n]^2 {
+	// Create qureg0 in superposition
+	x := 0:uint[n];
+	for i in [0..n) { x[i] := H(x[i]); }
+
+	// Perform a CNOT on each qubit of qureg1
+	y := 0:uint[n];
+	for i in [0..n) {
+		if x[i] { y[i] := X(y[i]); }
+	}
```

Comment:
This is of course a valid way to do it, but note that you can just write `y := dup(x);` for the same overall effect.

Reply:
Well isn't that more like entangling the states to form a bell pair/ghz state? I'm not sure how is that can be seen as teleportation?

Reply:
Great example; however please correct me, but isn't this is an example of an implementation of a Hadamard Gate -> more info https://en.wikipedia.org/wiki/Quantum_logic_gate#Hadamard_(H)_gate

Teleportation requires creating an array with an entangled state; then separating the array while in an entangled state, creating a separate quantum circuit. See this recent article for more information https://www.sciencealert.com/physicists-have-teleported-information-between-particles-of-matter-for-the-first-time

Reply:
> Well isn't that more like entangling the states to form a bell pair/ghz state? I'm not sure how is that can be seen as teleportation?

i agree.

Reply:
Re:
> Great example; however please correct me, but isn't this is an example of an implementation of a Hadamard Gate -> more info https://en.wikipedia.org/wiki/Quantum_logic_gate#Hadamard_(H)_gate
> 
> Teleportation requires creating an array with an entangled state; then separating the array while in an entangled state, creating a separate quantum circuit. See this recent article for more information https://www.sciencealert.com/physicists-have-teleported-information-between-particles-of-matter-for-the-first-time

Hi @ZoeDreams,
This is not an example of the implementation of the Hadamard gate. The H gate is used to create an equal superposition of one or more states something like this: H|0⟩ = 1/√2⋅(|0⟩ + |1⟩) and H|1⟩ = 1/√2⋅(|0⟩ - |1⟩). A similar thing can be extended to multiple qubits.

Yes, you're right. Teleportation requires creating an array with an entangled state; then separating the array while in an entangled state, creating a separate quantum circuit. And that's what is happening here. Do remember that quantum teleportation is also an algorithm and it is implemented here as an 'algorithm'. The function `createBellPair()` returns the two entangled registers `x` and `y`, which are then separated by passing one of them to the `alice()` function and another to the `bob()` function. The link which you mention is just the implementation of teleportation on two different physical processors, which is slightly different from what is being done here.

Here, teleportation is shown between two 'quregisters' belonging to the **same quantum system**. Whereas, in the experiment which has been mentioned by you, these two quregisters (electrons) exist as **two different quantum particles** within the same quantum system. Also, the communication channel between Alice and Bob is simulated by the logic that the output of `alice()` function is passed to `bob()` function, which then returns the teleported state.

One more thing is that this code just provides an algorithmic approach of Quantum Teleportation. The `main()` function will be the one to prepare the message quantum state to be teleported and also communicate the measured classical bits from Alice to Bob. That can be implemented as given [here](https://github.com/AbeerVaishnav13/Quantum-Algorithms-in-Silq/blob/master/Quantum-Algorithms/Teleportation.slq).

Reply:
Thank you for your reply and clarification. if i understand you in clear terms this is teleporting registers (pointers) to entangled bits in the same array? Again thank you for taking the time to respond clearly.